src/tsdb.c

```diff
@@ -15,52 +15,95 @@
 
 #include <inttypes.h>
 #include <math.h>
+#include <stdlib.h>
 #include "rmutil/alloc.h"
 #include "rmutil/logging.h"
 #include "rmutil/strings.h"
 
-static Series *lastDeletedSeries = NULL;
 static RedisModuleString *renameFromKey = NULL;
 
+static void deleteReferenceToDeletedSeries(Series *series,
+                                           SERIES_RELATION relation,
+                                           RedisModuleString *keyName) {
+    switch (relation) {
+        case SERIES_RELATION_DST:
+            SeriesDeleteRule(series, keyName);
+            break;
+        case SERIES_RELATION_SRC:
+            SeriesDeleteSrcRule(series, keyName);
+            break;
+        case SERIES_RELATION_NO_RELATION:
+            break;
+    }
+}
+
+CompactionRule *GetRule(CompactionRule *rules, RedisModuleString *keyName) {
+    CompactionRule *rule = rules;
+    while (rule != NULL) {
+        if (RedisModule_StringCompare(rule->destKey, keyName) == 0) {
+            return rule;
+        }
+        rule = rule->nextRule;
+    }
+    return NULL;
+}
+
 int GetSeries(RedisModuleCtx *ctx,
+              Series *originalSeries,
               RedisModuleString *keyName,
+              SERIES_RELATION relation,
```

Comment:
Lets move from an enum to a boolean: `shouldCleanRelation` and inside `deleteReferenceToDeletedSeries` open all the rules and subsequent keys and check if the references are broken. if it is clean srcKey or delete the rule.